core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -0,0 +1,319 @@
+          // with the transactional id at the same time; in this case we will
+          // treat it as the metadata has existed and update it accordingly
+          metadata synchronized {
+            if (!metadata.equals(newMetadata))
+              metadata.epoch = (metadata.epoch + 1).toShort
+          }
+
+          responseCallback(initTransactionMetadata(metadata))
+
+        case Some(metadata) =>
+          metadata synchronized {
+            metadata.epoch = (metadata.epoch + 1).toShort
+          }
+          responseCallback(initTransactionMetadata(metadata))
+      }
+    }
+  }
+
+  private def validateTransactionalId(transactionalId: String): Errors =
+    if (transactionalId == null || transactionalId.isEmpty)
+      Errors.INVALID_REQUEST
+    else if (!txnManager.isCoordinatorFor(transactionalId))
+      Errors.NOT_COORDINATOR
+    else if (txnManager.isCoordinatorLoadingInProgress(transactionalId))
+      Errors.COORDINATOR_LOAD_IN_PROGRESS
+    else
+      Errors.NONE
+
+
+  def handleAddPartitionsToTransaction(transactionalId: String,
+                                       pid: Long,
+                                       epoch: Short,
+                                       partitions: collection.Set[TopicPartition],
+                                       responseCallback: TxnMetadataUpdateCallback): Unit = {
+    val errors = validateTransactionalId(transactionalId)
+    if (errors != Errors.NONE)
+      responseCallback(errors)
+    else {
+      // try to update the transaction metadata and append the updated metadata to txn log;
+      // if there is no such metadata treat it as invalid pid mapping error.
+      val (error, newMetadata) = txnManager.getTransaction(transactionalId) match {
+        case None =>
+          (Errors.INVALID_PID_MAPPING, null)
+
+        case Some(metadata) =>
+          // generate the new transaction metadata with added partitions
+          metadata synchronized {
+            if (metadata.pid != pid) {
+              (Errors.INVALID_PID_MAPPING, null)
+            } else if (metadata.epoch != epoch) {
+              (Errors.INVALID_PRODUCER_EPOCH, null)
+            } else if (metadata.pendingState.isDefined) {
+              // return a retriable exception to let the client backoff and retry
+              (Errors.COORDINATOR_LOAD_IN_PROGRESS, null)
+            } else if (!TransactionMetadata.isValidTransition(metadata.state, Ongoing)) {
+              (Errors.INVALID_TXN_STATE, null)
+            } else if (partitions.subsetOf(metadata.topicPartitions)) {
+              // this is an optimization: if the partitions are already in the metadata reply OK immediately
+              (Errors.NONE, null)
+            } else {
+              val newMetadata = new TransactionMetadata(pid,
+                epoch,
+                metadata.txnTimeoutMs,
+                Ongoing,
+                metadata.topicPartitions ++ partitions,
+                time.milliseconds())
```

Comment:
Jun raises a good point. Right we should have two timestamps in the messages we write to the transaction log. One is the `entryTimestamp` which is the current time and which is used to expire the transactionalId. 

The other is the `transactionStartTime` which is the timestamp of the first `AddPartitionsToTransaction` request for a transaction. This is used to expire transactions on the broker.

As it stands, we have only the `entryTimestamp` and so a transaction could be open for ever if it gets updates very slowly. This would hold up the progress of the LSO on all the partitions involved in the transaction, which is not a desirable outcome.

@dguy would you be able to add the new timestamp and add a note to update doc on this ticket? https://confluentinc.atlassian.net/browse/CPKAFKA-616